subsys/net/lib/http/http_client.c

```diff
@@ -392,13 +392,60 @@ static void http_client_init_parser(struct http_parser *parser,
 	settings->on_url = on_url;
 }
 
-static int http_wait_data(int sock, struct http_request *req)
+static int http_wait_data(int sock, struct http_request *req, int32_t timeout)
 {
 	int total_received = 0;
 	size_t offset = 0;
 	int received, ret;
+	struct zsock_pollfd fds[1];
+	int nfds = 1;
+	int32_t remaining_time = timeout;
+	int64_t now = k_uptime_get();
+
+	fds[0].fd = sock;
+	fds[0].events = POLLIN;
 
 	do {
+		if (timeout > 0) {
+			remaining_time -=  (int32_t)k_uptime_delta(&now);
+			if (remaining_time < 0) {
+				/* timeout, make poll return immediately */
+				remaining_time = 0;
+			}
+		}
+
+		ret = zsock_poll(fds, nfds, remaining_time);
+		if (ret  == 0) {
+			LOG_DBG("Timeout");
+			ret = total_received;
+
+			if (req->internal.response.cb) {
+				NET_DBG("Calling callback for Timed out connection "
+					"(NULL HTTP response)");
+
+				/* Status code 0 representing a null response */
+				req->internal.response.http_status_code = 0;
+
+				/* Zero out related response metrics */
+				req->internal.response.processed = 0;
+				req->internal.response.data_len = 0;
+				req->internal.response.content_length = 0;
+				req->internal.response.body_frag_start = NULL;
+				memset(req->internal.response.http_status, 0,
+				       HTTP_STATUS_STR_SIZE);
+
+				req->internal.response.cb(&req->internal.response,
+							  HTTP_DATA_FINAL,
+							  req->internal.user_data);
+			}
+			break;
+		} else if (ret < 0) {
+			/* Socket error */
+			LOG_DBG("Connection error (%d)", errno);
+			ret = -errno;
+			break;
+		}
+
```

Comment:
What I miss here is analyzing the actual socket monitoring result. I. e. proceed only if `fds[0].revents & ZSOCK_POLLIN` and report an error in case of `fds[0].revents & ZSOCK_POLLERR` or `fds[0].revents & ZSOCK_POLLNVAL`. `ZSOCK_POLLHUP` could be treated the same was as `(received == 0)`

Reply:
@rlubos @wcappelle Sorry for a reply in an old PR, but i am having issues with it. Are you sure that 

> ZSOCK_POLLHUP could be treated the same was as (received == 0)

Right now I am receiving POLLHUP while the socket still has unprocessed data, which also aligns with POLLHUP's description.

```
       POLLHUP
              Hang up (only returned in revents; ignored in events).
              Note that when reading from a channel such as a pipe or a
              stream socket, this event merely indicates that the peer
              closed its end of the channel.  Subsequent reads from the
              channel will return 0 (end of file) only after all
              outstanding data in the channel has been consumed.
```

Reply:
I think you're right, replied in https://github.com/zephyrproject-rtos/zephyr/discussions/63030